### PR TITLE
Add Java bindings for string literal support in AST

### PR DESCRIPTION
**Pull request by karthikeyann**
## Description
Depends on #13061
Add Java bindings for string scalar support in AST
Add unit test for string comparison - column vs column, column vs literal.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: java/src/main/java/ai/rapids/cudf/ast/Literal.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ast/Literal.java:1]: 2023 copyrights

File: java/src/main/java/ai/rapids/cudf/ast/Literal.java [hunk 1]
```diff
@@ -200,6 +200,15 @@ public static Literal ofDurationFromLong(DType type, Long value) {
     return ofDurationFromLong(type, value.longValue());
   }
 
+  /** Construct a string literal with the specified value or null. */
+  public static Literal ofString(String value) {
+    byte[] stringBytes = value.getBytes();
```
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ast/Literal.java:205]: This doesn't handle `value` being null.  Need to do something like this:
```suggestion
    if (value == null) {
      return ofNull(DType.STRING)
    }
    byte[] stringBytes = value.getBytes();
```

File: java/src/test/java/ai/rapids/cudf/ast/CompiledExpressionTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/ast/CompiledExpressionTest.java:1]: 2023 copyrights

File: java/src/test/java/ai/rapids/cudf/ast/CompiledExpressionTest.java [hunk 2]
```diff
@@ -476,6 +476,65 @@ void testBinaryComparisonOperationTransform(BinaryOperator op, Integer[] in1, In
...
+        Arguments.of(BinaryOperator.GREATER, in1, in2, mapArray(in1, (a) -> a.compareTo(in2) > 0)),
+        Arguments.of(BinaryOperator.LESS_EQUAL, in1, in2, mapArray(in1, (a) -> a.compareTo(in2) <= 0)),
+        Arguments.of(BinaryOperator.GREATER_EQUAL, in1, in2, mapArray(in1, (a) -> a.compareTo(in2) >= 0)));
+  }
+
+  @ParameterizedTest
+  @MethodSource("createStringLiteralComparisonParams")
+  void testStringLiteralComparison(BinaryOperator op, String[] in1, String in2,
+                                               List<Boolean> expectedValues) {
+    Literal lit = Literal.ofString(in2);
```
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/ast/CompiledExpressionTest.java:496]: This doesn't test a null string literal, should add that as a test case.

I also recommend adding something like the `test*LiteralTransform` tests which are more focused on just handling the literal directly without other expressions and validates that the string data was translated properly.  The binary comparison test can hide a bad string translation and still pass the test by chance.

Comment (karthikeyann) [reply java/src/test/java/ai/rapids/cudf/ast/CompiledExpressionTest.java:496]: For string column, direct transformation is not supported. only fixed width types are supported now.

**Review by jlowe — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/ast/Literal.java [hunk 3]
```diff
@@ -200,6 +200,18 @@ public static Literal ofDurationFromLong(DType type, Long value) {
...
   }
 
+  /** Construct a string literal with the specified value or null. */
+  public static Literal ofString(String value) {
+    if (value == null) {
+      return ofNull(DType.STRING);
+    }
+    byte[] stringBytes = value.getBytes();
+    byte[] serializedValue = new byte[stringBytes.length + Integer.BYTES];
+    ByteBuffer.wrap(serializedValue).order(ByteOrder.nativeOrder()).putInt(stringBytes.length);
```
Comment (mythrocks) [inline java/src/main/java/ai/rapids/cudf/ast/Literal.java:210]: :+1:

File: java/src/main/native/src/CompiledExpression.cpp [hunk 4]
```diff
@@ -254,9 +263,29 @@ struct make_literal {
                                      std::move(scalar_ptr));
   }
 
+  /** Construct an AST literal from a string value */
+  template <typename T, std::enable_if_t<std::is_same_v<T, cudf::string_view>> * = nullptr>
```
Comment (mythrocks) [inline java/src/main/native/src/CompiledExpression.cpp:267]: I will look into JNI switching to using `CUDF_ENABLE_IF()`, afterwards.

File: java/src/main/native/src/CompiledExpression.cpp [hunk 5]
```diff
@@ -56,12 +57,20 @@ class jni_serialized_ast {
...
   /** Read a multi-byte value from the serialized AST data buffer */
   template <typename T> T read() {
-    check_for_eof(sizeof(T));
-    // use memcpy since data may be misaligned
-    T result;
-    memcpy(reinterpret_cast<jbyte *>(&result), data_ptr, sizeof(T));
-    data_ptr += sizeof(T);
-    return result;
+    if constexpr (std::is_same_v<T, std::string>) {
+      auto const size = read<cudf::size_type>();
```
Comment (mythrocks) [inline java/src/main/native/src/CompiledExpression.cpp:61]: :+1:

**Review by mythrocks — APPROVED** [review]
Awesome. LGTM!

**Pull request merged**